File: crates/citadel-mcp/src/bin/mcp.rs
//! `citadeldb-mcp`: a stdio MCP server exposing a memory region as tools.
//!
//! An MCP client (e.g. Claude Desktop) spawns this binary and talks JSON-RPC
//! over stdin/stdout; the passphrase is `CITADEL_KEY`, non-protocol output
//! goes to stderr. The region is encrypted by default (per-atom sealed + crypto
//! erasure); pass `--region-mode plaintext` to opt out.
//!
//! Every server invocation selects an embedder explicitly. The Candle embedder ships
//! in the default build; models are fetched only on request, never
//! automatically. Use `--embedder mock` only when keyword-only recall is
//! intentional:
//!
//! ```text
//! citadeldb-mcp pull e5-large              # download to ~/.citadel/models
//! citadeldb-mcp pull ms-marco-minilm       # cross-encoder reranker
//! citadeldb-mcp --db memory.cdl --embedder e5-large --reranker ms-marco-minilm
//! ```
//!
//! `e5-large` + `ms-marco-minilm` is the highest-recall config. Or
//! `--model-dir <dir>` for a compatible local checkpoint; a `cuda-embed`
//! build uses an NVIDIA GPU.
//!
//! Claude Desktop config (claude_desktop_config.json):
//! ```json
//! { "mcpServers": { "citadel": {
//!     "command": "citadeldb-mcp",
//!     "args": ["--db", "/path/memory.cdl", "--region", "default",
//!              "--embedder", "e5-large", "--reranker", "ms-marco-minilm"],
//!     "env": { "CITADEL_KEY": "<passphrase>" } } } }
//! ```

use std::process::ExitCode;

fn main() -> ExitCode {
    let argv: Vec<String> = std::env::args().skip(1).collect();
    match citadel_mcp::run(&argv) {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("citadeldb-mcp: {e}");
            ExitCode::FAILURE
        }
    }
}

File: python/citadeldb/mcp.py
"""MCP server: expose a citadel memory region as MCP tools over JSON-RPC stdio.

This wheel is built without the Candle embedder, so callers must explicitly pass
``embedder="mock"`` for keyword-only recall. Use the standalone
``citadeldb-mcp`` package (``uvx citadeldb-mcp``) for semantic recall. The
memory API accepts your own embeddings in either build.
"""

import sys
from collections.abc import Sequence

from citadeldb._core import mcp_main

__all__ = ["main", "serve", "pull"]


def main(argv: Sequence[str] | None = None) -> int:
    """Run the MCP CLI; returns the process exit code."""
    args = list(sys.argv[1:]) if argv is None else list(argv)
    return mcp_main(args)


def serve(
    db: str,
    *,
    embedder: str,
    region: str = "default",
    encrypted: bool = True,
    model_dir: str | None = None,
    models_dir: str | None = None,
    reranker: str | None = None,
    reranker_dir: str | None = None,
) -> int:
    """Serve with the explicitly selected embedder until the client disconnects."""
    args = [
        "--db", db,
        "--region", region,
        "--region-mode", "encrypted" if encrypted else "plaintext",
        "--embedder", embedder,
    ]
    if model_dir is not None:
        args += ["--model-dir", model_dir]
    if models_dir is not None:
        args += ["--models-dir", models_dir]
    if reranker is not None:
        args += ["--reranker", reranker]
    if reranker_dir is not None:
        args += ["--reranker-dir", reranker_dir]
    return mcp_main(args)


def pull(name: str, models_dir: str | None = None) -> int:
    """Download a public model into the local cache (needs a candle-embed build)."""
    args = ["pull", name]
    if models_dir is not None:
        args += ["--models-dir", models_dir]
    return mcp_main(args)


if __name__ == "__main__":
    sys.exit(main())
